File: nashorn/src/main/resources/JSON.js
(function (JSON) {
  var JsonArray = Java.type('io.vertx.core.json.JsonArray');
  var JsonObject = Java.type('io.vertx.core.json.JsonObject');

  var Map = Java.type('java.util.Map');
  var List = Java.type('java.util.List');

  // this will wrap the original function to handle Vert.x native types too
  var _stringify = JSON.stringify;
  var _parse = JSON.parse;

  // patch the original JSON object
  JSON.stringify = function () {
    var val = arguments[0];
    if (val instanceof JsonArray || val instanceof JsonObject) {
      return val.encode();
    }
    // convert from map to object
    if (val instanceof Map) {
      return new JsonObject(val).encode();
    }
    // convert from list to array
    if (val instanceof List) {
      return new JsonArray(val).encode();
    }
    return _stringify.apply(JSON, Array.prototype.slice.call(arguments))
  };

  // patch the original JSON object
  JSON.parse = function () {
    var val = arguments[0];
    if (val instanceof JsonArray) {
      return val.getList();
    }
    if (val instanceof JsonObject) {
      return val.getMap();
    }

    return _parse.apply(JSON, Array.prototype.slice.call(arguments))
  };

  // enhancement that converts json to java native objects
  JSON.native = function (arg) {
    if (arg === null) {
      return null;
    }

    var nativeJSON = Java.asJSONCompatible(arg);

    if (nativeJSON instanceof List) {
      return new JsonArray(nativeJSON);
    }

    if (val instanceof Map) {
      return new JsonObject(nativeJSON);
    }
  };
})(JSON);

File: factories/jython/src/main/java/xyz/jetdrone/vertx/JSON.java
package xyz.jetdrone.vertx;

import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.python.core.PyDictionary;
import org.python.core.PyList;

public final class JSON {

  private JSON() {}

  public static JsonObject to_native(PyDictionary dict) {
    if (dict == null) {
      return null;
    }
    return new JsonObject(dict);
  }

  public static JsonArray to_native(PyList list) {
    if (list == null) {
      return null;
    }
    return new JsonArray(list);
  }

  public static PyDictionary parse(JsonObject json) {
    if (json == null) {
      return null;
    }
    PyDictionary dict = new PyDictionary();
    dict.putAll(json.getMap());
    return dict;
  }

  public static PyList parse(JsonArray json) {
    if (json == null) {
      return null;
    }
    return new PyList(json.getList());
  }

  public static String stringify(PyDictionary dict) {
    if (dict == null) {
      return null;
    }

    return to_native(dict).encode();
  }

  public static String stringify(PyList list) {
    if (list == null) {
      return null;
    }

    return to_native(list).encode();
  }

  public static String stringify(JsonObject dict) {
    if (dict == null) {
      return null;
    }

    return dict.encode();
  }

  public static String stringify(JsonArray list) {
    if (list == null) {
      return null;
    }

    return list.encode();
  }
}
